src/emit/solana/mod.rs

```diff
@@ -1127,4 +1128,252 @@ impl SolanaTarget {
             }
         }
     }
+
+    /// Construct the LLVM-IR to call 'external_call' from solana.c
+    fn build_external_call<'b>(
+        &self,
+        binary: &Binary,
+        function: FunctionValue<'b>,
+        address: PointerValue<'b>,
+        payload: PointerValue<'b>,
+        payload_len: IntValue<'b>,
+        contract_args: ContractArgs<'b>,
+        ns: &ast::Namespace,
+        error_message: String,
+        loc: pt::Loc,
+    ) {
+        let parameters = self.sol_parameters(binary);
+        let external_call = binary.module.get_function("external_call").unwrap();
+
+        let program_id = contract_args.program_id.unwrap_or_else(|| {
+            binary
+                .llvm_type(&Type::Address(false), ns)
+                .ptr_type(AddressSpace::default())
+                .const_null()
+        });
+
+        let (seeds, seeds_len) = contract_args
+            .seeds
+            .map(|(seeds, len)| {
+                (
+                    seeds,
+                    binary.builder.build_int_cast(
+                        len,
+                        external_call.get_type().get_param_types()[5].into_int_type(),
+                        "len",
+                    ),
+                )
+            })
+            .unwrap_or((
+                external_call.get_type().get_param_types()[4]
+                    .ptr_type(AddressSpace::default())
+                    .const_null(),
+                external_call.get_type().get_param_types()[5]
+                    .into_int_type()
+                    .const_zero(),
+            ));
+
+        let ret = binary
+            .builder
+            .build_call(
+                external_call,
+                &[
+                    payload.into(),
+                    payload_len.into(),
+                    address.into(),
+                    program_id.into(),
+                    seeds.into(),
+                    seeds_len.into(),
+                    parameters.into(),
+                ],
+                "",
+            )
+            .try_as_basic_value()
+            .left()
+            .unwrap()
+            .into_int_value();
+
+        self.return_error(function, binary, ret, error_message, ns, loc);
```

Comment:
Maybe `external_call()` should be re-written to also abort on error (i.e. call `sol_panic()`).

Reply:
We should panic if `sol_invoke_signed_c` also panics, right?

Reply:
if `sol_invoke_signed_c` fails, the execution has already been aborted

Reply:
> Maybe `external_call()` should be re-written to also abort on error (i.e. call `sol_panic()`).

Done. A few tests had to be adjusted to accommodate the change.